### PR TITLE
feat: add new event hashing algorithm

### LINKED ISSUE #14259: Migration event hashing to new algorithm
Use the new hashing algorithm for new events. Hash old events according to the old algorithm. Put the new hashing algorithm for new events behind a feature flag such that it is only used for new events when the flag is enabled.

This ticket must not be completed until performance testing is complete. The performance tests do not need to "pass", but there must be results that inform the next steps of the project. The test must have high event volume, therefore it must be on a large network. Either perf1 or engnet2.

### PR DESCRIPTION
**Pull request by lpetrovic05**
closes #14259

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java [hunk 1]
```diff
@@ -16,16 +16,39 @@
...
+    private final boolean migrateEventHashing;
+
+    /**
+     * Constructs a new {@link DefaultEventHasher} with the given {@link SemanticVersion} and migration flag.
+     *
+     * @param currentSoftwareVersion the current software version
+     * @param migrateEventHashing    if true then use the new event hashing algorithm for new events, events created by
+     *                               previous software versions will still need to be hashed using the old algorithm.
+     */
+    public DefaultEventHasher(final SemanticVersion currentSoftwareVersion, final boolean migrateEventHashing) {
```
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java:38]: null annotation (and maybe null checks) missing

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java [hunk 2]
```diff
@@ -16,16 +16,39 @@
...
+        this.currentSoftwareVersion = currentSoftwareVersion;
+        this.migrateEventHashing = migrateEventHashing;
+    }
+
     @Override
     @NonNull
     public PlatformEvent hashEvent(@NonNull final PlatformEvent event) {
+        if (migrateEventHashing
```
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java:46]: null check of event missing

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/PbjHasher.java [hunk 3]
```diff
@@ -22,24 +22,37 @@
...
+public class PbjHasher implements EventHasher, UnsignedEventHasher {
 
     /** The hashing stream for the event. */
     private final HashingOutputStream eventHashingStream = new HashingOutputStream(DigestType.SHA_384.buildDigest());
     /** The hashing stream for the payloads. */
     private final HashingOutputStream payloadHashingStream = new HashingOutputStream(DigestType.SHA_384.buildDigest());
 
     @Override
+    @NonNull
     public PlatformEvent hashEvent(@NonNull final PlatformEvent event) {
```
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/PbjHasher.java:42]: null check of event missing

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/UnsignedEventHasher.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+package com.swirlds.platform.event.hashing;
+
+import com.swirlds.platform.system.events.UnsignedEvent;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Hashes unsigned events.
+ */
+public interface UnsignedEventHasher {
+    /**
```
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/UnsignedEventHasher.java:26]: can you add a newline before the comment starts?

Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/UnsignedEventHasher.java:26]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/PbjHasher.java (hunk 3 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/PbjHasher.java:42]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java (hunk 2 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java:46]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java (hunk 1 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/hashing/DefaultEventHasher.java:38]: added

**Review by litt3 — APPROVED** [review]

**Pull request merged**